sklearn/manifold/_graph_spectral_embedding.py

```diff
@@ -0,0 +1,676 @@
+from ..utils import check_symmetric
+from ..utils.extmath import randomized_svd
+
+def _augment_diagonal(graph, weight=1):
+    r"""
+    Replaces the diagonal of an adjacency matrix with :math:`\frac{d}{nverts - 1}` where
+    :math:`d` is the degree vector for an unweighted graph and the sum of magnitude of
+    edge weights for each node for a weighted graph. For a directed graph the in/out
+    :math:`d` is averaged.
+    Parameters
+    ----------
+    graph: {ndarray, sparse matrix}
+        Input graph in any of the above specified formats.
+
+    weight: float/int
+        scalar value to multiply the new diagonal vector by
+
+    Returns
+    -------
+    graph : np.array
+        Adjacency matrix with average degrees added to the diagonal.
+    """
+    dia = diags(graph.diagonal()) if isspmatrix_csr(graph) else np.diag(np.diag(graph))
+    graph = graph - dia
+
+    divisor = graph.shape[0] - 1
+
+    in_degrees = np.squeeze(np.asarray(abs(graph).sum(axis=0)))
+    out_degrees = np.squeeze(np.asarray(abs(graph).sum(axis=1)))
+
+    degrees = (in_degrees + out_degrees) / 2
+    diag = weight * degrees / divisor
+
+    graph += diags(diag) if isspmatrix_csr(graph) else np.diag(diag)
+
+    return graph
+
+def _selectSVD(X, n_components=None, n_elbows=2, svd_solver="randomized"):
+    r"""
+    Dimensionality reduction using SVD.
+    Performs linear dimensionality reduction by using either full singular
+    value decomposition (SVD) or truncated SVD. Full SVD is performed using
+    SciPy's wrapper for ARPACK, while truncated SVD is performed using either
+    SciPy's wrapper for LAPACK or Sklearn's implementation of randomized SVD.
+    It also performs optimal dimensionality selection using Zhu & Godsie svd_solver
+    if number of target dimension is not specified.
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        The data to perform svd on.
+    n_components : int or None, default = None
+        Desired dimensionality of output data. If "full",
+        ``n_components`` must be ``<= min(X.shape)``. Otherwise, ``n_components`` must be
+        ``< min(X.shape)``. If None, then optimal dimensions will be chosen by
+        :func:`~graspologic.embed.select_dimension` using ``n_elbows`` argument.
+    n_elbows : int, optional, default: 2
+        If ``n_components`` is None, then compute the optimal embedding dimension using
+        :func:`~graspologic.embed.select_dimension`. Otherwise, ignored.
+    svd_solver : {'randomized' (default), 'full', 'truncated'}, optional
+        SVD solver to use:
+        - 'randomized'
+            Computes randomized svd using
+            :func:`sklearn.utils.extmath.randomized_svd`
+        - 'full'
+            Computes full svd using :func:`scipy.linalg.svd`
+            Does not support ``graph`` input of type scipy.sparse.csr_matrix
+        - 'truncated'
+            Computes truncated svd using :func:`scipy.sparse.linalg.svds`
```

Comment:
```suggestion

    Parameters
    ----------
    X : array-like, shape (n_samples, n_features)
        The data to perform whose dimension are to reduce.
        
    n_components : int or None, default=None
        Desired dimensionality of output data. If "full", 
        ``n_components`` must be ``<= min(X.shape[0])``.
        Otherwise, ``n_components`` must be ``< min(X.shape[0])``.
        If `None`, then optimal dimensions will be chosen by
        :func:`~graspologic.embed.select_dimension` using ``n_elbows`` argument.

    n_elbows : int, optional, default=2
        If ``n_components`` is None, then compute the optimal embedding dimension using
        :func:`~graspologic.embed.select_dimension`. Otherwise, ignored.

    svd_solver : {'randomized' (default), 'full', 'truncated'}, optional
        SVD solver to use:
        - 'randomized'
            Computes randomized SVD using
            :func:`sklearn.utils.extmath.randomized_svd`
        - 'full'
            Computes full SVD using :func:`scipy.linalg.svd`
            Does not support ``graph`` input of type `scipy.sparse.csr_matrix`
        - 'truncated'
            Computes truncated SVD using :func:`scipy.sparse.linalg.svds`
```